arch/xtensa/core/xtensa_mmu.c

```diff
@@ -805,9 +809,80 @@ void arch_mem_unmap(void *addr, size_t size)
 		va += KB(4);
 	}
 
+#if CONFIG_MP_MAX_NUM_CPUS > 1
+	z_xtensa_mmu_tlb_ipi();
+#endif
+
 	k_spin_unlock(&xtensa_mmu_lock, key);
 }
 
+/* This should be implemented in the SoC layer.
+ * This weak version is here to avoid build errors.
+ */
+void __weak z_xtensa_mmu_tlb_ipi(void)
+{
+}
+
+void z_xtensa_mmu_tlb_shootdown(void)
```

Comment:
Mentioned this in the more specific comment below, but honestly our implementation of shootdown on x86 has always felt racy to me.  I know this has been a perenial source of bugs on Linux too.  I know know for sure that this is wrong, it's just too hard for that to fit in my brain at once.  Maybe a cleaner architecture would be to e.g. refcount the memory domain in the scheduler such that we can know that there's another CPU running on these page tables, and if there is come up with some lockout protocol via an IPI such that we know both threads are suspended/preempted while the operation happens?

Reply:
That sounds reasonable, and we can look try this out, I just want to avoid this change now, this code is being used / tested for some time and it is consistently working including stress tests. My proposal is to iterate over this later and try to align other platforms, as x86, too.